Source/Pawnmorphs/Esoteria/Genebank/ChamberDatabase.cs

```diff
@@ -0,0 +1,423 @@
+        public int TotalStorage
+        {
+            get => _totalStorage;
+            set => _totalStorage = Mathf.Max(0, value);
+        }
+
+        /// <summary>
+        ///     Gets the amount of storage space currently in use.
+        /// </summary>
+        /// <value>
+        ///     The used storage.
+        /// </value>
+        public int UsedStorage
+        {
+            get
+            {
+                if (_usedStorageCache == null)
+                {
+                    var v = 0;
+
+                    foreach (IList<IGenebankEntry> genebankEntries in _genebankDatabase.Values)
+                    {
+                        v += genebankEntries.Sum(x => x.GetRequiredStorage());
+                    }
+
+					_usedStorageCache = v;
+                }
+
+                return _usedStorageCache.Value;
+            }
+        }
+
+
+
+        /// <summary>
+        /// Gets a readonly list of values contained in genebank by type.
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <returns>Empty collection if no values.</returns>
+        public IReadOnlyList<T> GetEntryValues<T>()
+        {
+            Type entryType = typeof(GenebankEntry<T>);
+
+			if (_genebankDatabase.ContainsKey(entryType))
+            {
+			    return _genebankDatabase[entryType].Select(x => (x as GenebankEntry<T>).Value).ToList();
+            }
+            return new List<T>();
+		}
+
+        /// <summary>
+        /// Gets a readonly list of <see cref="GenebankEntry{T}"/> from genebank by type.
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <returns>Empty collection if no values.</returns>
+        public IReadOnlyList<GenebankEntry<T>> GetEntryItems<T>()
+		{
+			Type entryType = typeof(GenebankEntry<T>);
+
+			if (_genebankDatabase.ContainsKey(entryType))
+			{
+				return _genebankDatabase[entryType].Cast<GenebankEntry<T>>().ToList();
+			}
+            return new List<GenebankEntry<T>>();
+		}
+
+		/// <summary>
+		///     Gets a value indicating whether this instance can tag.
+		/// </summary>
+		/// <value>
+		///     <c>true</c> if this instance can tag; otherwise, <c>false</c>.
+		/// </value>
+		public bool CanTag => FreeStorage > 0;
+
+		/// <summary>
+		/// Adds the template to the database
+		/// </summary>
+		/// <param name="entry">The entry to add.</param>
+		/// <param name="failMode">The fail mode.</param>
+		/// <exception cref="ArgumentNullException">mutationDef</exception>
+		/// Note: this does
+		/// <b>not</b>
+		/// check if there is enough space to add the mutation or if it is restricted, use
+		/// <see cref="CanAddToDatabase(Pawnmorph.Hediffs.MutationDef)" />
+		/// to check
+		public bool AddToDatabase<T>([NotNull] GenebankEntry<T> entry, LogFailMode failMode = LogFailMode.Silent)
+		{
+			if (entry == null)
+				throw new ArgumentNullException(nameof(entry));
+
+            if (CanAddToDatabase(entry, out string reason) == false)
```

Comment:
FYI, TryAddToDatabase calls CanAddToDatabase as well as this method, which means CanAddToDatabase gets called twice.  Based on the docstring of this method, this method wasn't intended to do any error checking, so this check could be removed

Reply:
This is correct, however it serves 2 different intentions.

TryAdd calls CanAdd to get reasons if it cannot be added to show.

However this one calls it to ensure data integrity, to prevent duplicates and overfilling the datacapacity.

Without it, you rely on caller to ALWAYS call CanAdd before calling this, which opens for circumventing logic and making mistakes.

I agree that it isn't optimal, and in my opinion they should all be just a single method instead that returns reason if it fails.
The reason i didn't change that now was in case another mod relies on the Add function not having additional parameters.

There might also be other limitations on why something shouldn't be addable, like chaos.

Reply:
I may have made some incorrect statements in the last reply since it was based on memory and i will revise it when I can access the code properly.

Reply:
I didn't actually notice the docstring saying that it contains no validation.

What are the reasons to have a method to circumvent our "business logic" here?

I try to aim to prevent future me from making mistakes by accident, and it seems to me like this is asking for someone to at some point make a mistake and result in adding invalid entries.

Saving and loading circumvent it because that is handled internally. So it doesn't matter if capacity is dead when saving and loading.

Reply:
I don't know what the original reason was.  I can't imagine many situations where it would be necessary, I agree that it's bad form.  It might be worth deprecating the method and moving the "unprotected add" logic into a private method instead.